### PR TITLE
Evelyn #u01 view map

### PR DESCRIPTION
**Pull request by evelynfoy**
# Business Analysis 🟠 Code Institute 🟠 Community Project - Pull Request 🎉🎉

## Pull Request References:
- #3 

### Summary
- [x] The index.html template displaying a map of Dún Laoghaire–Rathdown
- [x]  A marker for each company in the organisations variable passed to the template
- [x] Marker Clusters for grouping markers
- [x] The company list is currently populated from the test file 'demo_data.json'.
- [ ]  The company list will need to be populated from the database once available.

### Points of note
1) The company list will need to be populated from the database once available. I have included a commented-out line that should work to retrieve the data from the database but this is untested.
2) api.mapbox.com requires an access token to work - see details below
3) The list of companies is made available in script.js using an inline script in base.html. This results in the list being visible in developer tools when it runs. I am not sure if there is a better way to do this.

### Details 

- A new template called index.html    

- A new style sheet called style.css in static/css containing styling for the leaflet container

- Two new stylesheets required to implement marker clusters using leafletjs.
   - MarkerCluster.Default.css in static/css
   - MarkerCluster.css in in static/css
 
- A new script required to implement marker clusters using leafletjs.
  -  leaflet.markercluster-src.js in static/js
- Updates to script.js in static/js
1. Add map to div on index.html with id = 'map'
2. Add a marker to a markerClusterGroup for each company in the companies list
3. Add the markerClusterGroup to the map on index.html
4. api.mapbox.com requires an accessToken to work. One can be requested but I used the one in the example as it worked ok  but we may need to request one
    - https://leafletjs.com/SlavaUkraini/examples/quick-start/example-basic.html
    - https://account.mapbox.com/access-tokens/
- Updates to app.py 
1. Change the @app.route("/") from def hello(): to def index():
2. Add line to load data from test file 'demo_data.json' into variable *organisations*
3. Add commented-out line to get data from database once available
4. Add line to return the rendered template index.html and pass the variable *organisations*

- Updates to the base.html template in templates
1) Add links to Leaflet stylesheets (3)
2) Add links to Leaflet scripts (2)
3) Add inline script to convert the JSON list of companies passed to the template to a list of companies which is then used by script.js to add markers for each company. 

## Breaking changes
  - This change will change the default path to display index.html rather than '"In the words of Théoden....<br>'So... it beings...'". If it does not work then the app could potentially crash but what it is currently doing isn't all that useful anyway.

## Screenshots/Gifs/Media
Home Page before clicking on a cluster
IndexBeforeClickingOnCluster
Home Page after clicking on a cluster
IndexAfterClickingOnCluster2
Mobile View
IndexAfterClickingOnCluster2Mobile

Where I got the access token  

mapBoxExampleAccessToken
'

### CODE REVIEW

File: static/css/MarkerCluster.Default.css [hunk 1]
```diff
@@ -0,0 +1,60 @@
+.marker-cluster-small {
```
Comment (stefdworschak) [inline static/css/MarkerCluster.Default.css:1]: I wonder about the file names. Could you use snake-case (all lowercase and underscores) or is this a pre-defined filename format that Leaflet uses?

File: static/js/scripts.js [hunk 2]
```diff
@@ -2,4 +2,34 @@
 var tooltipTriggerList = [].slice.call(document.querySelectorAll('[data-bs-toggle="tooltip"]'))
 var tooltipList = tooltipTriggerList.map(function (tooltipTriggerEl) {
   return new bootstrap.Tooltip(tooltipTriggerEl)
-})
\ No newline at end of file
+})
+
+// Add map to index.html page
+var map = L.map('map').setView([53.27, -6.20], 13);
```
Comment (stefdworschak) [inline static/js/scripts.js:8]: could you maybe wrap your code in a function as well?

File: app.py [hunk 3]
```diff
@@ -19,9 +19,11 @@
 
 
 @app.route("/")
-def hello():
-    return "In the words of Théoden....<br>'So... it beings...'"
-
+def index():
+    organisations = json.load(open('demo_data.json'))
```
Comment (stefdworschak) [inline app.py:23]: Could you please revert this change? You should be able to insert a few organisations in the mongo database and load them on the frontend the exact same way you are doing right now as well. If you need help with this please let me know.

**Review by stefdworschak — CHANGES_REQUESTED** [review]
Some small changes here. The main thing is to go back to loading the organisations from the mongodb instead of the json file directly. Please let me know if you need any help with anything and when I can re-review the Pull Request.

**Comment by stefdworschak** [comment]
Also, there are a few merge-conflicts that will have to be resolved before merging. Let me know if you need help with these.

File: static/js/scripts.js (hunk 2 above)
Comment (auxfuse) [reply static/js/scripts.js:8]: @stefdworschak I'm going to resolve this for now and we can continue to refactor and add some reusable functions in due course. 🤜

File: static/css/style.css [hunk 4]
```diff
@@ -0,0 +1,7 @@
+.leaflet-container {
+    height: 85vh;
+    width: 100%;
+    max-width: 100%;
+    max-height: 100%;
+    margin-top: 2rem;
+}
```
Comment (auxfuse) [inline static/css/style.css:7]: Newline at End of File needed here.

**Review by auxfuse — CHANGES_REQUESTED** [review]
Just a couple of changes, and once done we can merge! Thanks for all your hard work on this @evelynfoy Fantastic inclusion.

File: static/css/style.css (hunk 4 above)
Comment (auxfuse) [reply static/css/style.css]: Resolved.

File: app/static/css/MarkerCluster.css [hunk 5]
```diff
@@ -0,0 +1,14 @@
+.leaflet-cluster-anim .leaflet-marker-icon, .leaflet-cluster-anim .leaflet-marker-shadow {
+	-webkit-transition: -webkit-transform 0.3s ease-out, opacity 0.3s ease-in;
```
[the remark this answers is not in the document]
Comment (auxfuse) [reply app/static/css/MarkerCluster.css:2]: Resolved.

File: app/static/css/MarkerCluster.Default.css [hunk 6]
```diff
@@ -0,0 +1,60 @@
+.marker-cluster-small {
+	background-color: rgba(181, 226, 140, 0.6);
```
[the remark this answers is not in the document]
Comment (auxfuse) [reply app/static/css/MarkerCluster.Default.css:2]: Resolved.

**Review by auxfuse — APPROVED** [review]
LGTM 🎉

**Review by stefdworschak — APPROVED** [review]
LGTM
